### PR TITLE
DataSketches extension

### PR DESCRIPTION
**Pull request by jtigani**
- Dummy sketch extension that just adds a clone of the SUM() aggregation. I don't really know what I'm doing, hopefully this is only visible to me. :-/
- Added a simple function that appends 'foo' to a string
- Basic hll init working
- Add missing files. Still figuring out git
- Add other integer types.
- Revert "Add other integer types."
- Still figuring out git. Add untracked hll files
- Add support for strings & decimals
- Add support for hll_count_extract function, which lets you get cardinality estimate from the sketch
- Fixed up memory usage
- Add support for hll_count_merge to merge multiple sketches
- Remove dummy functions
- minor formatting tweaks

### CODE REVIEW

**Comment by Mause** [comment]
I assume you didn't mean to submit this pr to the duckdb repo?

**Review by Mytherin — COMMENTED** [review]
Hey Jordan :) Thanks for the draft PR!

In general the code looks good to me, some minor comments below:

**Comment by tdoehmen** [comment]
Sorry to jump in here. Just wanted to say I'm glad to see this PR! I remember talking to @Mytherin about such an integration a while back, but sadly never got around to doing it. 

Are you planning to add integrations with all the different sketch types offered in apache-datasketches, or are you mainly interested in the distinct counting sketches? 

An KLL and Count-Min sketches integration would be really nice to have for an upcoming project. If it helps, I'd be happy to work on those.

File: Makefile [hunk 1]
```diff
@@ -136,132 +136,2415 @@ ifneq ($(BUILD_OUT_OF_TREE_EXTENSION),)
 	EXTENSIONS:=${EXTENSIONS} -DEXTERNAL_EXTENSION_DIRECTORY=$(BUILD_OUT_OF_TREE_EXTENSION)
 endif
 
+# Default target executed when no arguments are given to make.
```
[the remark this answers is not in the document]
Comment (jtigani) [reply Makefile]: Fixed ... looks like I still need to fix formatting.  Still wrestling with clang-tidy.

File: extension/sketch/sketch-hll.cpp [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+            // Switch from adding values to unioning sketches mode.
+            this->union_sketch->update(*this->value_sketch);
+            delete(this->value_sketch);
+            this->value_sketch = nullptr;
+        }
+        return *this->union_sketch;
+    }
+    
+  private:
+    mutable datasketches::hll_sketch* value_sketch;
```
[the remark this answers is not in the document]
Comment (jtigani) [reply extension/sketch/sketch-hll.cpp:152]: It doesn't look like the constructor gets called on these state objects, which leaves unique_ptr objects in a bad state.

**Comment by jtigani** [comment]
Hi @tdoehmen, I was planning to work my way through the various sketch families... I just started with HLL because they are pretty straightforward and I'm familiar with how they are exposed in BigQuery.  I would be happy to work with you on the other ones, or you can just wait for me to get around to them :-)

File: Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (jtigani) [reply Makefile]: (fixed clang-tidy)

File: extension/sketch/sketch-hll.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jtigani) [reply extension/sketch/sketch-hll.cpp]: I based it on https://github.com/duckdb/duckdb/blob/master/src/function/aggregate/distributive/approx_count.cpp which also uses bare new / delete. I can create a new class to encapsulate those three objects and can be properly constructed when needed, but that seems a bit ocercomplicated. Or if you think there should be a way to make the unique_ptr work, I'll figure something out.

**Comment by tdoehmen** [comment]
Sounds great! I forked your branch and you can basically just throw at me whatever sketch you'd like to "outsource" :) It's nice to have a reference for the HLL function names like BigQuery. I wonder if it make sense to look at datasketches-postgresql [0] as a reference for the how to expose the other sketches. OTOH the SQL function names there seem a bit long and clucky. 

[0] https://github.com/apache/datasketches-postgresql/tree/master/test

File: extension/sketch/sketch-hll.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Mytherin) [reply extension/sketch/sketch-hll.cpp]: Indeed, that is correct, I missed that. It's fine as it is, although there are some places where the new/delete is not exception safe and an exception will cause memory leaks to happen.

**Pull request closed without merging**